### ISSUE TITLE
Allow a group's commitee to backup a group's files and pages

### ISSUE THREAD

**Issue by gerdneuman** [riseuplabs/crabgrass-core#236]
During the last two years it was unclear to us, if crabgrass would be continued. So we were fearful of using a maybe discontinued software (deadend)... At the same time having all our group's files and documents on a "foreign" server (that might once be seized or shutdown) and no backup leaves us still a bit "unsettled"

Would be great if there was a way to backup all of a group's content easily. Just having everything in an zip-archive would be fine (simple html as it is seen in the browser), not really database dump. We already thought of crawling all our files using something like wget or [ScrabBook](https://addons.mozilla.org/de/firefox/addon/scrapbook/), but we were afraid that these tools might hit the "Delete" links within crabgrass and therefore accidentically remove all our content.

**Comment by elijh** [comment]
We do hope to support this. I can promise that we will either continue to develop crabgrass or we will add this feature!

**Comment by meitar** [comment]
I would like to second this feature request. Even if Crabgrass does continue to be developed, I would like to se this feature implemented. The ability to export one's dataset makes using the hosted Crabgrass instance much more approachable because it means that one's data is not forever in some format that is only meaningfully accessible if the remote service is available.

Further, I work in some environments where the absense of a local backup precludes the use of Crabgrass by policy. Adding this feature would help me make the case to certain groups to move to Crabgrass away from less safe and less philosophically aligned alternatives such as Gogle Docs.

See also the many, many dupes of this feature request:

* https://labs.riseup.net/code/issues/475
* https://labs.riseup.net/code/issues/754
* https://labs.riseup.net/code/issues/1270
* https://labs.riseup.net/code/issues/1702
* https://labs.riseup.net/code/issues/4099
* https://labs.riseup.net/code/issues/4419

TL;DR: Please implememt this feature regardless of commitments to keep we.riseup.net alive. It is an important data portability issue and software libre issue on its own merit. Thank you!

**Comment by azul** [comment]
So my first take on this would be a crawler script that zips the result.
You would end up with an archive of html files, css files and images in a directory layout similar to the crabgrass urls.
This could be browsed locally (at least in parts) - but it would not expose the internal data structure.

What's the key thing here? Being able to import the data somewhere else or being able to browse it?

**Comment by meitar** [comment]
> What's the key thing here? Being able to import the data somewhere else or being able to browse it?

Since Crabgrass itself is free software, I think the ideal result of this issue is the implementation of an export format that is re-importable to another Crabgrass instance.

I can, of course, use `wget` or an equivalent to `--mirror` a given group, as long as I am assured the "Delete" button won't accidentally be pressed, as @gerdneuman notes in the original issue. But creating such a static backup doesn't provide the ability to pick up working where we left off; a lot of post-processing would have to be done to take the data from a static mirror and massage it into a format that is editable in the same or an equivalent way as it was on Crabgrass. For instance, I'd have to write another script to parse the HTML of the wiki pages and insert them into new MediaWiki pages, or something.

What is really needed is something more like [WordPress's XWR](https://codex.wordpress.org/Tools_Export_Screen) format that exports the content but not the internal structure of a particular blog, so it is easy to move content from one instance of the WordPress software to another, independent of the underlying database.

**Comment by gerdneuman** [comment]
@azul: A locally browsable archive containing just of files/html/css/images would be fine for our group and practically enough. We have several documents (wiki pages, minutes, scanned legal documents) that need to be saved someway.Thanks :heart: 

I see the features that @meitar proposes as two separate issues:

- a) export the db structure so I can setup my own crabcrass and 
- b) export to some kind of export format like XWR . 

I think these have definitely their merits but more so for the more technically educated. Most people do not know how to setup their own crabcrass... These should be handled in different tickets. They are also technically much different each I assume.

**Comment by meitar** [comment]
> A locally browsable archive containing just of files/html/css/images would be fine for our group and practically enough. We have several documents (wiki pages, minutes, scanned legal documents) that need to be saved someway.Thanks

@gerdneuman, does this mean that if @riseuplabs provided a kind of "snapshot my group" tool as part of the Crabgrass distribution, that would satisfy your group's need?

I can probably whip up a script to do a static snapshot over HTTP from a command line (`bash`?) for a specific group and have a first draft done by tonight. However, if you are suggesting that you would need this functionality exposed as a button from within the Crabgrass group's interface itself, that's a different story.

**Comment by gerdneuman** [comment]
@meitar: A script would be totally fine for us. Since it is a script we could even run it via cron on a weekly basis to backup :+1:  For not so geeky users a button would probably be easier to use.

**Comment by meitar** [comment]
> A script would be totally fine for us.

Okay, I am "bored" tonight, let me see what I can do in short order.

**Comment by meitar** [comment]
Okay, I think I have a super simple script for you to try out and give some feedback on if you're willing, @gerdneuman. Have a look at https://gist.github.com/meitar/30ba639017192025cc01 and let me know what you think.

**Comment by meitar** [comment]
Hey @gerdneuman @azul, sorry I've been slow to respond to your comments on the gist I linked. Apparently GitHub doesn't send me notifications when you leave comments on the gist (see isaacs/github#21). I didn't know this when I made that gist. If you have comments about it you want me to see, please ping me about them here even if you leave a comment there so that I will be notified about them.

**Comment by gerdneuman** [comment]
@meitar I left a comment https://gist.github.com/meitar/30ba639017192025cc01#gistcomment-1685626 Really a pity that github does not send notifications for gists, thx for letting me know!

**Comment by meitar** [comment]
@gerdneuman, saw your comment, and I see what you mean. The folks I work with don't tend to create committees often, and when they do it isn't often intended to be a long-lived group. But I understand your use case and I've modified the script to add a `--subgroup` option that will automatically detect and mirror any committees or councils that it can find.

Have another go and let me know what you think?

**Comment by meitar** [comment]
Just FYI, I decided I hate gists so I made [a repo](https://github.com/meitar/crabgrass-tools) to house the backup script.

**Comment by gerdneuman** [comment]
I close this as a simple backup can now be made using https://github.com/meitar/crabgrass-tools

I'll create extra tickets for the db dump idea and the export idea.
